weejson-jsoniter-scala/src/main/scala/com/rallyhealth/weejson/v1/wee_jsoniter_scala/WeePickleJsonValueCodecs.scala

```diff
@@ -0,0 +1,199 @@
+        if (depthM1 < 0) in.decodeError("depth limit exceeded")
+        val isEmpty = in.isNextToken('}')
+        val obj = v.visitObject(if (isEmpty) 0 else -1).narrow
+        if (!isEmpty) {
+          in.rollbackToken()
+          var key: String = "?"
+          try {
+            while ({
+              key = in.readKeyAsString()
+              obj.visitKeyValue(obj.visitKey().visitString(key))
+              obj.visitValue(decodeValue(in, obj.subVisitor, depthM1))
+              in.isNextToken(',')
+            }) ()
+            if (!in.isCurrentToken('}')) in.objectEndOrCommaError()
+          } catch {
+            case NonFatal(t) => prependFailurePathInfo(t, key)
+          }
+        }
+        obj.visitEnd()
+      } else {
+        in.readNullOrError(v.visitNull(), "expected `null` value")
+      }
+    }
+
+    private def parseNumber[J](
+      in: JsonReader,
+      v: Visitor[_, J]
+    ): J = {
+      in.setMark()
+      var digits = 0
+      var b = in.nextByte()
+      if (b == '-') b = in.nextByte()
+      try {
+        while (b >= '0' && b <= '9') {
+          b = in.nextByte()
+          digits += 1
+        }
+      } catch {
+        case _: JsonReaderException => // ignore the end of input error for now
+      } finally in.rollbackToMark()
+
+      if ((b | 0x20) != 'e' && b != '.') {
+        if (digits < 19) {
+          val l = in.readLong()
+          v.visitInt64(l)
+        } else {
+          val x = in.readBigInt(null)
+          if (x.bitLength < 64) v.visitInt64(x.longValue)
+          else v.visitFloat64StringParts(x.toString, -1, -1)
+        }
+      } else {
+        val cs = new String(in.readRawValAsBytes(), StandardCharsets.US_ASCII)
+
+        /**
+          * This regex performs rather badly, but gets the tests passing.
+          *
+          * We're looking for a value we can pass through the Visitor interface--
+          * either a primitive Double or a CharSequence representing a *valid*
+          * number conforming to https://datatracker.ietf.org/doc/html/rfc7159#page-6.
+          *
+          * `in.readRawValAsBytes()` does NOT do that validation. It will happily
+          * return a String of "------".
+          *
+          * `in.readBigDecimal(null).toString` is tempting, but will not provide the raw input.
+          * Instead, it transforms the input from "0.00000001" to "1.0E-8".
+          * This fails roundtrip tests.
+          *
+          * I tried combining the two approaches, `in.readBigDecimal(null)` for validation,
+          * then `in.rollbackToMark()` + `in.readRawValAsBytes()` to capture the raw input,
+          * but for a value like "1.0-----", `in.readBigDecimal(null)` will read "1.0",
+          * then `in.readRawValAsBytes()` will return the whole string, including the unwanted
+          * trailing hyphens.
+          *
+          */
+        require(ValidJsonNum.pattern.matcher(cs).matches(), "invalid number")
+        v.visitFloat64String(cs)
```

Comment:
@plokhotnyuk, I seem to be stuck here. Do you have any suggestions to achieve better performance than my hacky regex given the failed attempts in the comment above? Specifically, is there anything in `JsonReader` that can help me that I might have overlooked?

Reply:
@htmldoug You can use some counters to calculate number of valid chars like in the following code snippet (BEWARE: it is not tested)
```
    private def parseNumber[J](
      in: JsonReader,
      v: Visitor[_, J]
    ): J = {
      in.setMark()
      var b = in.nextByte()
      var digits, index = 0
      var decIndex, expIndex = -1
      if (b == '-') {
        b = in.nextByte()
        index += 1
      }
      try {
        digits -= index
        while (b >= '0' && b <= '9') {
          b = in.nextByte()
          index += 1
        }
        digits += index
        if (b == '.') {
          decIndex = index
          b = in.nextByte()
          index += 1
        }
        digits -= index
        while (b >= '0' && b <= '9') {
          b = in.nextByte()
          index += 1
        }
        digits += index
        if ((b | 0x20) == 'e') {
          expIndex = index
          b = in.nextByte()
          index += 1
        }
        while (b >= '0' && b <= '9') {
          b = in.nextByte()
          index += 1
        }
      } catch {
        case _: JsonReaderException => // ignore the end of input error for now
      } finally in.rollbackToMark()
      if ((decIndex & expIndex) == -1) {
        if (digits < 19) v.visitInt64(in.readLong())
        else {
          val x = in.readBigInt(null)
          if (x.bitLength < 64) v.visitInt64(x.longValue)
          else v.visitFloat64StringParts(x.toString, -1, -1)
        }
      } else {
        val cs = new String(in.readRawValAsBytes(), StandardCharsets.US_ASCII)
        require(cs.length - 1 == index, "invalid number")
        v.visitFloat64String(cs)
      }
    }
```

Reply:
@htmldoug You can parse an array of numbers to make your benchmark more realistic. I bet nobody use JSON parsers to read a single number.